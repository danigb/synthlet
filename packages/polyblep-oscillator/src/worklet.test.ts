import { getProcessorName } from "./index";

describe("ProcessorNode", () => {
  let Processor: any;
  const sampleRate = 40;

  beforeAll(async () => {
    createWorkletTestContext(sampleRate);
    Processor = (await import("./worklet")).Processor;
  });

  it("registers processor", () => {
    expect(global.registerProcessor).toHaveBeenCalledWith(
      getProcessorName(),
      Processor
    );
  });

  it("has parameter descriptors", () => {
    expect(Processor.parameterDescriptors).toMatchSnapshot();
  });

  it("generates sine", () => {
    const processor = new Processor();
    const { inputs, outputs } = createInputsOutputs({ length: sampleRate });
    const params = {
      waveform: [0],
      frequency: [2],
    };
    processor.process(inputs, outputs, params);
    expect(outputs).toMatchSnapshot();
  });
  it("generates saw", () => {
    const processor = new Processor();
    const { inputs, outputs } = createInputsOutputs({ length: sampleRate });
    const params = {
      waveform: [1],
      frequency: [2],
    };
    processor.process(inputs, outputs, params);
    expect(outputs).toMatchSnapshot();
  });
  it("generates square", () => {
    const processor = new Processor();
    const { inputs, outputs } = createInputsOutputs({ length: sampleRate });
    const params = {
      waveform: [2],
      frequency: [2],
    };
    processor.process(inputs, outputs, params);
    expect(outputs).toMatchSnapshot();
  });
  it("generates triangle", () => {
    const processor = new Processor();
    const { inputs, outputs } = createInputsOutputs({ length: sampleRate });
    const params = {
      waveform: [3],
      frequency: [2],
    };
    processor.process(inputs, outputs, params);
    expect(outputs).toMatchSnapshot();
  });
});

function createWorkletTestContext(sampleRate = 10) {
  // @ts-ignore
  global.sampleRate = sampleRate;
  // @ts-ignore
  global.AudioWorkletProcessor = class AudioWorkletNodeStub {
    port: {
      postMessage: jest.Mock<any, any, any>;
      onmessage: jest.Mock<any, any, any>;
    };

    constructor() {
      this.port = {
        postMessage: jest.fn(),
        onmessage: jest.fn(),
      };
    }
  };
  // @ts-ignore
  global.registerProcessor = jest.fn(); // Mock registerProcessor
}

function createInputsOutputs(
  options: { ins?: number; outs?: number; length?: number } = {}
) {
  const inCount = options.ins ?? 1;
  const outCount = options.outs ?? 1;
  const length = options.length ?? 10;
  const inputs: Float32Array[][] = [];
  const outputs: Float32Array[][] = [];

  for (let i = 0; i < inCount; i++) {
    inputs.push([new Float32Array(length)]);
  }
  for (let i = 0; i < outCount; i++) {
    outputs.push([new Float32Array(length)]);
  }
  return { inputs, outputs };
}
