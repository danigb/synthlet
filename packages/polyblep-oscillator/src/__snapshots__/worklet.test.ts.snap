// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessorNode generates saw 1`] = `
[
  [
    Float32Array [
      0,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -2,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      -8.881784197001252e-15,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      -1,
      8.881784197001252e-15,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
      1,
    ],
  ],
]
`;

exports[`ProcessorNode generates sine 1`] = `
[
  [
    Float32Array [
      0,
      -0.8999999761581421,
      -0.800000011920929,
      -0.699999988079071,
      -0.6000000238418579,
      -0.5,
      -0.4000000059604645,
      -0.30000001192092896,
      -0.20000000298023224,
      -0.10000000149011612,
      -1.1102230246251565e-16,
      0.10000000149011612,
      0.20000000298023224,
      0.30000001192092896,
      0.4000000059604645,
      0.5,
      0.6000000238418579,
      0.699999988079071,
      0.800000011920929,
      0.8999999761581421,
      -8.43769498715119e-15,
      -0.8999999761581421,
      -0.800000011920929,
      -0.699999988079071,
      -0.6000000238418579,
      -0.5,
      -0.4000000059604645,
      -0.30000001192092896,
      -0.20000000298023224,
      -0.10000000149011612,
      4.440892098500626e-16,
      0.10000000149011612,
      0.20000000298023224,
      0.30000001192092896,
      0.4000000059604645,
      0.5,
      0.6000000238418579,
      0.699999988079071,
      0.800000011920929,
      0.8999999761581421,
    ],
  ],
]
`;

exports[`ProcessorNode generates square 1`] = `
[
  [
    Float32Array [
      0,
      -1.600000023841858,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -3.2000725269317627,
      1.5998547077178955,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      0.00007264318264788017,
      -1.600000023841858,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -1.6000726222991943,
      -0.00007264318264788017,
      1.600000023841858,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
      1.6000726222991943,
    ],
  ],
]
`;

exports[`ProcessorNode generates triangle 1`] = `
[
  [
    Float32Array [
      0,
      -0.8999999761581421,
      -0.800000011920929,
      -0.699999988079071,
      -0.6000000238418579,
      -0.5,
      -0.4000000059604645,
      -0.30000001192092896,
      -0.20000000298023224,
      -0.10000000149011612,
      -1.1102230246251565e-16,
      0.10000000149011612,
      0.20000000298023224,
      0.30000001192092896,
      0.4000000059604645,
      0.5,
      0.6000000238418579,
      0.699999988079071,
      0.800000011920929,
      0.8999999761581421,
      -8.43769498715119e-15,
      -0.8999999761581421,
      -0.800000011920929,
      -0.699999988079071,
      -0.6000000238418579,
      -0.5,
      -0.4000000059604645,
      -0.30000001192092896,
      -0.20000000298023224,
      -0.10000000149011612,
      4.440892098500626e-16,
      0.10000000149011612,
      0.20000000298023224,
      0.30000001192092896,
      0.4000000059604645,
      0.5,
      0.6000000238418579,
      0.699999988079071,
      0.800000011920929,
      0.8999999761581421,
    ],
  ],
]
`;

exports[`ProcessorNode has parameter descriptors 1`] = `
[
  {
    "automationRate": "k-rate",
    "defaultValue": 0,
    "maxValue": 2,
    "minValue": 0,
    "name": "type",
  },
  {
    "automationRate": "k-rate",
    "defaultValue": 440,
    "maxValue": 20000,
    "minValue": 0,
    "name": "frequency",
  },
  {
    "automationRate": "k-rate",
    "defaultValue": 0,
    "maxValue": 10000,
    "minValue": 0,
    "name": "detune",
  },
]
`;
